src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts

```diff
@@ -867,7 +869,20 @@ export class TerminalTaskSystem implements ITaskSystem {
 				delete this.sameTaskTerminals[taskKey];
 			}
 		} else if (allowsSharedTerminal) {
-			let terminalId = this.idleTaskTerminals.remove(taskKey) || this.idleTaskTerminals.shift();
+			// Always allow to reuse the terminal previously used by the same task.
+			let terminalId = this.idleTaskTerminals.remove(taskKey);
+			if (!terminalId) {
+				// There is no idle terminal which was used by the same task.
+				// Search for any idle terminal used previously by a task of the same terminalGroup
+				// (or, if the task has no terminalGroup, a terminal used by a task without terminalGroup).
+				for (const taskId of this.idleTaskTerminals.keys()) {
```

Comment:
Can you add to the comment and explain what happens when there is not terminal group?